File: Codeforces/Div 3/Codeforces Round 981 (Div. 3)/F_Kosuke_s_Sloth.cpp
#include <bits/stdc++.h>
using namespace std;
using LL = long long;
const int MOD = 1e9 + 7;
#define endl '\n'

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    LL t, n, k;
    cin >> t;
    unordered_map<LL, tuple<LL, LL, vector<LL>>> pisano;
    pisano[1] = {1, 1, {1}};
    while (t--) {
        cin >> n >> k;
        if (pisano.find(k) == pisano.end()) {
            LL f_pre = 1, f_cur = 1;
            vector<LL> pos;
            LL i = 1;
            while (true) {
                if (f_pre % k == 0) {
                    pos.push_back(i);
                }
                LL tmp = f_pre;
                f_pre = f_cur;
                f_cur = (tmp + f_cur) % k;
                i++;
                if (f_pre == 1 && f_cur == 1) {
                    break;
                }
            }
            pisano[k] = {i-1, pos.size(), pos};
        }
        auto [ln, cnt, pos] = pisano[k];
        if (cnt == 0) {
            cout << -1 << endl;
            continue;
        }
        LL q = n / cnt;
        LL r = n % cnt;
        if (r == 0) {
            cout << (q * ln) % MOD << endl;
        } else {
            cout << ((q * ln) % MOD + pos[r-1]) % MOD << endl;
        }
    }
    return 0;
}

File: Codeforces/Div 3/Codeforces Round 981 (Div. 3)/F_Kosuke_s_Sloth.py
MOD = 10**9 + 7

pisano = {1: (1, 1, [1])}
t = int(input())
for _ in range(t):
    n, k = map(int, input().split())

    if k not in pisano:
        f_pre, f_cur = 1, 1
        pos = []
        i = 1
        while True:
            if f_pre % k == 0:
                pos.append(i)
            f_pre, f_cur = f_cur, (f_pre + f_cur) % k
            i += 1
            if f_pre == 1 and f_cur == 1:
                ln = i - 1
                break
        pisano[k] = (ln, len(pos), pos)

    ln, cnt, pos = pisano[k]
    if cnt == 0:
        print(-1)
        continue
    q, r = divmod(n, cnt)
    if r == 0:
        index = (q * ln) % MOD
    else:
        index = ((q * ln) + pos[r-1]) % MOD
    print(index)